mmdet3d/models/losses/lovasz_loss.py

```diff
@@ -0,0 +1,398 @@
+            B, H, W = probs.size()
+            probs = probs.view(B, 1, H, W)
+        B, C, H, W = probs.size()
+        probs = probs.permute(0, 2, 3, 1).contiguous().view(-1,
+                                                            C)  # B*H*W, C=P,C
+        labels = labels.view(-1)
+    if ignore_index is None:
+        return probs, labels
+    valid = (labels != ignore_index)
+    vprobs = probs[valid.nonzero().squeeze()]
+    vlabels = labels[valid]
+    return vprobs, vlabels
+
+
+def lovasz_hinge_flat(logits: torch.Tensor,
+                      labels: torch.Tensor) -> torch.Tensor:
+    """Binary Lovasz hinge loss.
+
+    Args:
+        logits (torch.Tensor): Logits at each prediction
+            (between -infty and +infty) with shape [P].
+        labels (torch.Tensor): Binary ground truth labels (0 or 1)
+            with shape [P].
+
+    Returns:
+        torch.Tensor: The calculated loss.
+    """
+    if len(labels) == 0:
+        # only void pixels, the gradients should be 0
+        return logits.sum() * 0.
+    signs = 2. * labels.float() - 1.
+    errors = (1. - logits * signs)
+    errors_sorted, perm = torch.sort(errors, dim=0, descending=True)
+    perm = perm.data
+    gt_sorted = labels[perm]
+    grad = lovasz_grad(gt_sorted)
+    loss = torch.dot(F.relu(errors_sorted), grad)
+    return loss
+
+
+def lovasz_hinge(logits: torch.Tensor,
+                 labels: torch.Tensor,
+                 classes: Optional[Union[str, List[int]]] = None,
+                 per_sample: bool = False,
+                 class_weight: Optional[List[float]] = None,
+                 reduction: str = 'mean',
+                 avg_factor: Optional[int] = None,
+                 ignore_index: int = 255) -> torch.Tensor:
+    """Binary Lovasz hinge loss.
+
+    Args:
+        logits (torch.Tensor): Logits at each pixel
+            (between -infty and +infty) with shape [B, H, W].
+        labels (torch.Tensor): Binary ground truth masks (0 or 1)
+            with shape [B, H, W].
+        classes (Union[str, list[int]], optional): Placeholder, to be
+            consistent with other loss. Defaults to None.
+        per_sample (bool): If per_sample is True, compute the loss per
+            sample instead of per batch. Defaults to False.
+        class_weight (list[float], optional): Placeholder, to be consistent
+            with other loss. Defaults to None.
+        reduction (str): The method used to reduce the loss. Options
+            are "none", "mean" and "sum". This parameter only works when
+            per_sample is True. Defaults to 'mean'.
+        avg_factor (int, optional): Average factor that is used to average
+            the loss. This parameter only works when per_sample is True.
+            Defaults to None.
+        ignore_index (Union[int, None]): The label index to be ignored.
+            Defaults to 255.
+
+    Returns:
+        torch.Tensor: The calculated loss.
```

Comment:
Will the shape of the loss change according to samplewise=True or False?

Reply:
No, the loss will be reduced in the end.